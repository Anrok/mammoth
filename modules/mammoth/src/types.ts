import { PickByValue } from 'utility-types';
import { ColumnWrapper } from './columns';

// These types are used to switch between optional and required properties. Because these are
// specific types it's possible to use TypeScript's condition types feature to map nullable columns
// to optional keys.
export interface Null<T> {
  nullType: T;
}

export interface NotNull<T> {
  notNullType: T;
}

export type toNotNull<T> = T extends Null<any> ? NotNull<T['nullType']> : NotNull<T>;
export type toNull<T> = T extends NotNull<any> ? Null<T['notNullType']> : Null<T>;
export type toType<T> = T extends NotNull<any>
  ? T['notNullType']
  : T extends Null<any>
  ? T['nullType']
  : T;

export interface SparseArray<T> {
  [index: number]: T | undefined;
}

export const getOrUndefined = <A>(array: SparseArray<A>, index: number): A | undefined =>
  array[index];

export type TypeOf<T> = {
  [K in keyof T]: toType<T[K]>;
};

export type Nullable<T> = {
  // To make sure explicitly passing null is allowed as well
  [P in keyof T]?: T[P] | null;
};

export type SplitOptionalAndRequired<
  Table extends { [columnName: string]: ColumnWrapper<any, any, any, any, any> },
  ColumnType extends 'insertType' | 'selectType' | 'updateType'
> = TypeOf<PickByValue<{ [K in keyof Table]: Table[K][ColumnType] }, NotNull<any>>> &
  Nullable<TypeOf<PickByValue<{ [K in keyof Table]: Table[K][ColumnType] }, Null<any>>>>;
