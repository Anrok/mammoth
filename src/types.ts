import { Column } from './column';
import { DeleteQuery } from './delete';
import { InsertQuery } from './insert';
import { NamedExpression } from './expression';
import { SelectQuery } from './select';
import { UpdateQuery } from './update';

export type ResultType = 'ROWS' | 'AFFECTED_COUNT';

export type PickByValue<T, ValueType> = Pick<
  T,
  {
    [Key in keyof T]-?: T[Key] extends ValueType ? Key : never;
  }[keyof T]
>;

export type Query =
  | SelectQuery<any>
  | UpdateQuery<any, any>
  | InsertQuery<any, any>
  | DeleteQuery<any, any>;

type GetColumnDataType<C> = C extends Column<any, any, infer DataType, infer IsNotNull, any, any>
  ? IsNotNull extends true
    ? DataType
    : DataType | undefined
  : never;

export type GetReturning<TableColumns, ColumnName extends keyof TableColumns> = {
  [K in ColumnName]: GetColumnDataType<TableColumns[K]>;
};

class DataType<Type, IsNull> {
  private _!: Type & IsNull;
}

export type ResultSet<T extends Query> = T extends SelectQuery<infer Selectables>
  ? {
      [K in keyof Selectables]: Selectables[K] extends Column<
        any,
        any,
        infer D,
        infer N,
        any,
        infer JoinType
      >
        ? Extract<JoinType, 'left-join'> extends never
          ? Extract<JoinType, 'left-side-of-right-join'> extends never
            ? Extract<JoinType, 'full-join'> extends never
              ? N extends true
                ? DataType<D, true>
                : DataType<D, false>
              : DataType<D, false>
            : DataType<D, false>
          : DataType<D, false>
        : Selectables[K] extends NamedExpression<any, infer D, infer IsNotNull>
        ? DataType<D, IsNotNull>
        : Selectables[K] extends SelectQuery<{}>
        ? ResultSet<Selectables[K]>[keyof ResultSet<Selectables[K]>]
        : never;
    }
  : T extends UpdateQuery<any, infer C>
  ? C
  : T extends InsertQuery<any, infer C>
  ? C
  : T extends DeleteQuery<any, infer C>
  ? C
  : never;
