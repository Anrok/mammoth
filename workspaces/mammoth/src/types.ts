export type ResultType = 'ROWS' | 'AFFECTED_COUNT';

export type PickByValue<T, ValueType> = Pick<
  T,
  {
    [Key in keyof T]-?: T[Key] extends ValueType ? Key : never;
  }[keyof T]
>;

export type GetReturning<TableColumns, ColumnName extends keyof TableColumns> = {
  [K in ColumnName]: TableColumns[K];
};

export class Maybe<T> {
  private _!: T;
}

// This is used to create a snapshot of the data type without erasing null or undefined. If the type
// is nullabe (not not null) the type is wrapped in a Maybe<T> helper class which is written in the
// snapshot as-is.
export type GetDataType<Type, IsNotNull extends boolean> = [IsNotNull] extends [true]
  ? [true] extends [IsNotNull]
    ? Type
    : Maybe<Type>
  : Maybe<Type>;

export type QueryExecutorFn = (
  query: string,
  parameters: any[],
) => Promise<{ rows: any[]; affectedCount: number }>;

export type Err<Message extends string> = {} & Message;
