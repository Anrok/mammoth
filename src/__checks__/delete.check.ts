import { Query, ResultSet } from '../types';
import { count, defineDb, defineTable, integer, text, timestampWithTimeZone, uuid } from '..';

/** @dts-jest enable:test-type */

const foo = defineTable(`foo`, {
  id: uuid().primaryKey().default(`gen_random_id()`),
  createDate: timestampWithTimeZone().notNull().default(`now()`),
  name: text().notNull(),
  value: integer(),
});

const toSnap = <T extends Query>(query: T): ResultSet<T> => {
  return undefined as any;
};

const db = defineDb(() => Promise.resolve({ rows: [], affectedRowsCount: 0 }));

// @dts-jest:group update
{
  // @dts-jest:snap should delete and returning id
  toSnap(db.deleteFrom(foo).returning(`id`));
}
