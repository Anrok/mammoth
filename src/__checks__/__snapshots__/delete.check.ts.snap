// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`update should delete and returning id (type) should match snapshot 1`] = `"GetReturning<{ id: Column<\\"id\\", \\"foo\\", string, true, any, undefined>; createDate: Column<\\"createDate\\", \\"foo\\", Date, true, any, undefined>; name: Column<\\"name\\", \\"foo\\", string, true, false, undefined>; value: Column<\\"value\\", \\"foo\\", number, false, false, undefined>; }, \\"id\\">"`;
